beacon-chain/forkchoice/doubly-linked-tree/store.go

```diff
@@ -84,21 +88,50 @@ func (s *Store) head(ctx context.Context) ([32]byte, error) {
 	if bestDescendant == nil {
 		bestDescendant = justifiedNode
 	}
-	currentEpoch := slots.EpochsSinceGenesis(time.Unix(int64(s.genesisTime), 0))
+	currentSlot := slots.SinceGenesis(time.Unix(int64(s.genesisTime), 0))
```

Comment:
The complexity of the head function is immense. I know we prefer to have big unit tests for its behavior, but it is becoming harder to deal with when we refactor. There are a lot of tricky edge-cases in here as well that are behind multiple conditionals and mathematical computations I am concerned about adding to an already big function

Reply:
> The way the proposer head is computed in the `head()` function itself is a side-effect. Instead, if it can be computed on the fly by an actual proposer when needed, it would reduce complexity of the head function and make it easier to write code that reads like the design doc for this pull request. Ideally, something like this, where `X` and `Y` are defined such as in the document https://www.notion.so/prysmaticlabs/Enabling-proposer-to-reorg-against-late-parent-block-55761f83c72442dd98d6db033122dfa3
> 
> ```
> func APIGetBlock() {
>    parentRoot := forkchoice.ComputeProposerHead()
> }
> 
> func computeProposerHead() {
> if seenBlockBeforeX() {
>   return normalLogic()
> }
> if inX {
>   return xIntervalLogic()
> }
> return yIntervalLogic()
> }
> ```
> 
> and each interval's logic for deciding the block the proposer will build on reads cleanly

This can be achieved only by increasing the memory footprint of forkchoice and I'm not convinced it will reduce the complexity. The only way I can now think how to implement this this way would be by keeping cached the number of votes that the late block had at `y`. I'll give it a shot and we can compare both approaches.

Reply:
Understood - I do think we should still simplify some of this logic with cleaner helper functions as it is hard to reason about in isolation. Specifically, the computations we use to decide how to set the proposer head